### PR TITLE
compute stats on gold answers

### PR DESCRIPTION
**Pull request by fiorda**

### CODE REVIEW

File: pybossa/stats/gold.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+            return
+        self.cm[true_ix][seen_ix] += 1
+
+
+def compute(stat, taskrun, gold, path):
+    return _compute(stat, Answer(taskrun), gold, path.split('.'))
+
+
+def _compute(stat, taskrun, gold, path_parts):
+    if isinstance(gold, list):
```
Comment (markroseberry) [inline pybossa/stats/gold.py]: You should separate the path iteration logic from the comparison logic. It is confusing to combine them both in one function.

File: pybossa/stats/gold.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+import logging
+
+import numpy as np
+
+
+logger = logging.getLogger(__name__)
+logger.addHandler(logging.NullHandler())
+
+
+class NoAnswer(object):
```
Comment (markroseberry) [inline pybossa/stats/gold.py]: I don't understand the intent behind the `Answer` and `NoAnswer` classes. It looks like you want to wrap the task results in a standard interface, but I don't see what your code is adding to the built-in `list` and `dict` types.

File: pybossa/stats/gold.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+def compute(stat, taskrun, gold, path):
+    return _compute(stat, Answer(taskrun), gold, path.split('.'))
+
+
+def _compute(stat, taskrun, gold, path_parts):
+    if isinstance(gold, list):
+        if not path_parts and stat.compare_lists:
+            return stat.update(taskrun, gold)
+        else:
+            for ans, gold_ans in zip(taskrun, gold):
```
Comment (markroseberry) [inline pybossa/stats/gold.py]: Do you have a test for this code? I think you don't or you would have seen the infinite iterators.

File: pybossa/stats/gold.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    _no_ans = NoAnswer()
+
+    def __init__(self, ans):
+        self.ans = ans
+
+    def __getitem__(self, key):
+        return Answer(self.ans.get(key, self._no_ans))
+
+    def __iter__(self):
+        if isinstance(self.ans, list):
```
[the remark this answers is not in the document]
Comment (fiorda) [reply pybossa/stats/gold.py]: ans is a json and can contain lists

File: pybossa/stats/gold.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+            base = self.ans
+        else:
+            base = []
+        return (Answer(x) for x in chain(base, self._no_ans))
+
+
+### Equalities
+
+
+def equality(one, two):
```
[the remark this answers is not in the document]
Comment (fiorda) [reply pybossa/stats/gold.py]: can do

File: pybossa/stats/gold.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    return one == two
+
+
+### Statistics
+
+
+class Statistic(object):
+
+    def update(self, seen, true):
```
[the remark this answers is not in the document]
Comment (fiorda) [reply pybossa/stats/gold.py]: can change

File: pybossa/stats/gold.py (hunk 1 above)
Comment (fiorda) [reply pybossa/stats/gold.py]: I can make a traversing generator function

**Pull request merged**